google/cloud/bigquery/dbapi/cursor.py

```diff
@@ -445,9 +459,61 @@ def _format_operation(operation, parameters=None):
             ``parameters`` argument.
     """
     if parameters is None or len(parameters) == 0:
-        return operation.replace("%%", "%")  # Still do percent de-escaping.
+        return operation.replace("%%", "%"), None  # Still do percent de-escaping.
+
+    operation, parameter_types = _extract_types(operation)
+    if parameter_types is None:
+        raise exceptions.ProgrammingError(
+            f"Parameters were provided, but {repr(operation)} has no placeholders."
+        )
 
     if isinstance(parameters, collections_abc.Mapping):
-        return _format_operation_dict(operation, parameters)
+        return _format_operation_dict(operation, parameters), parameter_types
+
+    return _format_operation_list(operation, parameters), parameter_types
+
+
+def _extract_types(
+    operation, extra_type_sub=re.compile(r"%(\(([^:)]*)(:(\w+))?\))?s").sub,
```

Comment:
Random thought - how to we treat strings like the following?
```
SELECT "%%(name:type)s" as weird_literal;
```

Percent-formatted strings treat double percent signs as a literal percent and would keep the following string intact:
```py
>>> 'SELECT "%%(name)s" as weird_literal;' % {"name": "foo"}
'SELECT "%(name)s" as weird_literal;'
```
Presumably our extension should follow suit for consistency, and would thus benefit from a negative lookbehind assertion?
```py
r"(?<!%)%(\(([^:)]*)(:(\w+))?\))?s"
```

Reply:
+1

Reply:
done